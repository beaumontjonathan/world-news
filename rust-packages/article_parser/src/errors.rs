use chrono::ParseError as ChronoParseError;
use htmlescape::DecodeErr;
use scraper::Selector;
use serde_json::Error as SerdeJsonError;

#[derive(Debug)]
pub enum InternalParseError {
  UnknownError(&'static str),
  HtmlStructureError(HtmlStructureError),
  JsonParseError(JsonParseError),
  InvalidDateTime(ChronoParseError),
  HtmlDecodeErr(DecodeErr),
}

#[derive(Debug)]
pub enum JsonParseErrorCause {
  InvalidLdJson,
  InvalidDataAttrJson,
}

#[derive(Debug)]
pub struct JsonParseError {
  pub cause: JsonParseErrorCause,
  pub serde_error: SerdeJsonError,
}

#[derive(Debug)]
pub enum HtmlErrorCause {
  MissingElement,
  NonUniqueElement,
  MissingAttribute(String),
}

#[derive(Debug)]
pub struct HtmlStructureError {
  pub cause: HtmlErrorCause,
  pub selector: &'static Selector,
}

impl InternalParseError {
  pub fn html(error: HtmlStructureError) -> InternalParseError {
    InternalParseError::HtmlStructureError(error)
  }

  pub fn json(error: JsonParseError) -> InternalParseError {
    InternalParseError::JsonParseError(error)
  }

  pub fn ld_json(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidLdJson,
      serde_error,
    })
  }

  pub fn json_data_attr(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidDataAttrJson,
      serde_error,
    })
  }

  pub fn unknown(s: &'static str) -> InternalParseError {
    InternalParseError::UnknownError(s)
  }

  pub fn chrono(err: ChronoParseError) -> InternalParseError {
    InternalParseError::InvalidDateTime(err)
  }

  pub fn decode_html(err: DecodeErr) -> InternalParseError {
    InternalParseError::HtmlDecodeErr(err)
  }
}

impl From<ChronoParseError> for InternalParseError {
  fn from(err: ChronoParseError) -> Self {
    InternalParseError::chrono(err)
  }
}

impl From<DecodeErr> for InternalParseError {
  fn from(err: DecodeErr) -> Self {
    InternalParseError::decode_html(err)
  }
}
