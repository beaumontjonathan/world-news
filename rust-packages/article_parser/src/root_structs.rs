use chrono::{DateTime, Utc};
use serde::Serialize;
use std::fmt::Debug;

#[derive(Debug, Serialize)]
pub struct Image {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Video {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Publisher {
  BBC,
  DailyMail,
  Independent,
  Metro,
  Mirror,
  Sky,
  Sun,
  Guardian,
}

use Publisher::*;

impl From<String> for Publisher {
  fn from(publisher: String) -> Self {
    match publisher.to_lowercase().as_str() {
      "bbc" => BBC,
      "dailymail" => DailyMail,
      "independent" => Independent,
      "metro" => Metro,
      "mirror" => Mirror,
      "sky" => Sky,
      "sun" => Sun,
      "theguardian" | "guardian" => Guardian,
      _ => panic!("Invalid publisher type"),
    }
  }
}

#[derive(Debug, Serialize)]
pub struct NewsArticleContent {
  pub headline: String,
  pub twitter_headline: Option<String>,
  pub description: String,
  pub thumbnail: Image,
  pub categories: Vec<String>,
  pub images: Vec<Image>,
  pub videos: Vec<Video>,
  pub body: String,
  pub date_updated: DateTime<Utc>,
  pub date_published: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct VideoArticleContent {
  pub something: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Serialize)]
pub enum ParsedPageContent {
  NewsArticle(NewsArticleContent),
  VideoArticle(VideoArticleContent),
}

#[derive(Debug, Serialize)]
pub struct PageMeta {
  pub publisher: Publisher,
  pub url: String,
  pub date_parsed: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ParsedPage {
  pub meta: PageMeta,
  pub content: ParsedPageContent,
}
