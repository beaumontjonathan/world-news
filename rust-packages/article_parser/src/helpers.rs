use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
// use chrono::{DateTime, Utc};
use crate::root_structs::ParsedPageContent;
use lazy_static::lazy_static;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};
// use std::fmt::{self, Debug, Display, Formatter};
pub trait ToOwnedOrNone {
  fn to_owned_or_none(self) -> Option<String>;
}

impl<'a> ToOwnedOrNone for Option<&'a str> {
  fn to_owned_or_none(self) -> Option<String> {
    self.and_then(|mut s| {
      s = s.trim();
      if s.is_empty() {
        None
      } else {
        Some(String::from(s))
      }
    })
  }
}
// pub trait NoneIfEmpty
// where
//   Self: Sized,
// {
//   fn none_if<F>(self, callback: F) -> Option<Self>
//   where
//     F: Fn(&Self) -> bool,
//   {
//     if callback(&self) {
//       None
//     } else {
//       Some(self)
//     }
//   }
// }
// impl<T> NoneIfEmpty for T {}

pub type ParseResult = Result<ParsedPageContent, InternalParseError>;

pub trait FixHtmlWhitespace {
  fn fix_html_whitespace(self) -> Self;
}

fn fix_whitespace(text: String) -> String {
  lazy_static! {
    static ref RE: Regex = Regex::new(r#"\s+"#).unwrap();
  }
  RE.replace_all(&text, " ").into_owned()
}

impl FixHtmlWhitespace for String {
  fn fix_html_whitespace(self) -> String {
    fix_whitespace(self)
  }
}

pub trait AttrValueOrNone {
  fn attr_value_or_none(&self, attr: &str) -> Option<String>;
}

impl<'a> AttrValueOrNone for ElementRef<'a> {
  fn attr_value_or_none(&self, attr: &str) -> Option<String> {
    self.value().attr(attr).to_owned_or_none()
  }
}

pub trait FullFixedHtmlText {
  fn full_fixed_html_text(self) -> String;
}

impl<'a> FullFixedHtmlText for ElementRef<'a> {
  fn full_fixed_html_text(self) -> String {
    self
      .text()
      .map(|text| text.trim().to_string())
      .collect::<Vec<String>>()
      .join(" ")
      .fix_html_whitespace()
      .trim()
      .to_owned()
  }
}

pub trait JoinWithNewline {
  fn join_with_newline(self) -> String;
}

impl<T> JoinWithNewline for T
where
  T: Iterator<Item = String>,
{
  fn join_with_newline(self) -> String {
    self.collect::<Vec<String>>().join("\n").trim().to_owned()
  }
}

pub trait UniqueElementRef<'a> {
  fn unique(self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause>;
}

impl<'a, T> UniqueElementRef<'a> for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn unique(mut self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause> {
    match (self.next(), self.next()) {
      (None, _) => Err(HtmlErrorCause::MissingElement),
      (_, Some(_)) => Err(HtmlErrorCause::NonUniqueElement),
      (Some(el_ref), None) => Ok(el_ref),
    }
  }
}

pub trait SelectHelpers {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError>;
}

impl SelectHelpers for Html {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

impl SelectHelpers for scraper::element_ref::ElementRef<'_> {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

pub trait FormattedText {
  fn formatted_text(self) -> String;
}

impl<'a, T> FormattedText for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn formatted_text(self) -> String {
    self.map(|p| p.full_fixed_html_text()).join_with_newline()
  }
}

pub trait ChildrenHelpers {
  fn children_first(&self) -> Option<ElementRef>;
  fn children_all(&self) -> Vec<ElementRef>;
}

impl ChildrenHelpers for scraper::element_ref::ElementRef<'_> {
  fn children_first(&self) -> Option<ElementRef> {
    self.children_all().into_iter().next()
  }

  fn children_all(&self) -> Vec<ElementRef> {
    self.children().filter_map(ElementRef::wrap).collect()
  }
}
