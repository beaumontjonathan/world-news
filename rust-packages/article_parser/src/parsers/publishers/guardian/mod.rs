use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
use crate::helpers::{
  AttrValueOrNone, ChildrenHelpers, FullFixedHtmlText, JoinWithNewline, ParseResult, SelectHelpers,
};
use crate::root_structs::{Image, NewsArticleContent, ParsedPageContent, Video};

use crate::my_selectors;
use chrono::{DateTime, Utc};
use scraper::{Html, Selector};

my_selectors! {
  ARTICLE_VIDEO_PRECHECK => "body > div > article[id=\"article\"][itemtype=\"http://schema.org/VideoObject\"].content--media--video",
  TWITTER_TITLE_SELECTOR => "head > meta[name=\"twitter:text:title\"]",
  DATE_UPDATED_SELECTOR => "meta[property=\"article:modified_time\"]",
  DATE_PUBLISHED_SELECTOR => "meta[property=\"article:published_time\"]",
  HEADER => "div.content__main-column > header",
  HEADLINE => "h1.content__headline",
  DESCRIPTION => "div.content__standfirst",
  FIGURE_HEADER => "figure.media-primary",
  IMG => "img",
  FIGCAPTION => "figcaption.caption",
  ARTICLE_CONTENT => "div.content__article-body",
  ELEMENT_IMAGE => "figure.element-image",
  ELEMENT_INTERACTIVE => "figure.element-interactive",
  VIDEO_FIGURE => "figure.element[data-atom-type=\"media\"]",
  YOUTUBE_IFRAME => "div.youtube-media-atom__iframe",
}

type Result<T> = std::result::Result<T, InternalParseError>;

pub fn parse(document: Html) -> ParseResult {
  // PRECHECKS
  match document.select(&ARTICLE_VIDEO_PRECHECK).next() {
    None => Ok(()),
    Some(_) => Err(InternalParseError::unknown(
      "Article type: video article, not yet supported",
    )),
    // Some(_) => Err(InternalParseError::precondition(
    //   "Article type: video article, not yet supported",
    // )),
  }?;

  let header = document.select_unique(&HEADER)?;
  let article_content = document.select_unique(&ARTICLE_CONTENT)?;

  // HEADLINE
  let headline = {
    let h1s = header
      .select(&HEADLINE)
      .map(|h1| h1.full_fixed_html_text())
      .collect::<Vec<_>>();
    match h1s.len() {
      0 => Err(InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector: &HEADLINE,
      })),
      1 => Ok(h1s.into_iter().next().unwrap()),
      _ => {
        if h1s.windows(2).all(|w| w[0] == w[1]) {
          Ok(h1s.into_iter().next().unwrap())
        } else {
          Err(InternalParseError::unknown(
            "Unknown headline type - too many <h1>s with differing text",
          ))
        }
      }
    }
  }?;

  // TWITTER HEADLINE
  let twitter_headline = document
    .select_unique_attr(&TWITTER_TITLE_SELECTOR, "content")
    .ok();

  // DESCRIPTION
  let description = header.select_unique(&DESCRIPTION)?.full_fixed_html_text();

  // THUMBNAIL
  // todo - implement
  let thumbnail = {
    let figure = header.select_unique(&FIGURE_HEADER)?;
    let img = figure.select_unique(&IMG)?;
    let alt = img.attr_value_or_none("alt");
    let url = img.attr_value_or_none("src");
    let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
    Image { alt, url, caption }
  };

  // CATEGORIES
  // TODO - implement
  let categories: Vec<String> = vec![];

  // IMAGES
  let images = article_content
    .children_all()
    .into_iter()
    .filter(|el| ELEMENT_IMAGE.matches(&el) || ELEMENT_INTERACTIVE.matches(&el))
    .map(|figure| {
      if ELEMENT_IMAGE.matches(&figure) {
        let img = figure.select_unique(&IMG)?;
        let alt = img.attr_value_or_none("alt");
        let url = img.attr_value_or_none("src");
        let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
        Ok(Image { alt, url, caption })
      } else if ELEMENT_INTERACTIVE.matches(&figure) {
        // unimplemented!();
        Ok(Image {
          alt: None,
          url: None,
          caption: None,
        })
      } else {
        panic!("how are you here?");
      }
    })
    .collect::<Result<Vec<_>>>()?;

  // VIDEOS
  let videos = article_content
    .children_all()
    .into_iter()
    .filter(|figure| VIDEO_FIGURE.matches(&figure))
    .map(|figure| {
      let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
      let youtube_id = figure.select_unique_attr(&YOUTUBE_IFRAME, "data-asset-id")?;
      let url = Some(format!("https://www.youtube.com/watch?v={}", youtube_id));

      Ok(Video {
        alt: None,
        url,
        caption,
      })
    })
    .collect::<Result<Vec<_>>>()?;

  // BODY
  let body = article_content
    .children_all()
    .into_iter()
    .filter_map(|el| {
      if el.value().name() == "p" || el.value().name() == "ul" {
        Some(el.full_fixed_html_text())
      } else {
        None
      }
    })
    .collect::<Vec<String>>()
    .into_iter()
    .join_with_newline();

  // DATE_UPDATED
  let date_updated = document.select_unique_attr(&DATE_UPDATED_SELECTOR, "content")?;
  let date_updated = DateTime::parse_from_rfc3339(&date_updated)?.with_timezone(&Utc);

  //DATE_PUBLISHED
  let date_published = document.select_unique_attr(&DATE_PUBLISHED_SELECTOR, "content")?;
  let date_published = DateTime::parse_from_rfc3339(&date_published)?.with_timezone(&Utc);

  Ok(ParsedPageContent::NewsArticle(NewsArticleContent {
    headline,
    twitter_headline,
    description,
    thumbnail,
    categories,
    images,
    videos,
    body,
    date_updated,
    date_published,
  }))
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::test_helpers::helpers::read_html_doc;
  use insta::assert_debug_snapshot;

  #[test]
  fn test_file_01() {
    let document = read_html_doc("test-01.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_02() {
    let document = read_html_doc("test-02.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_03() {
    let document = read_html_doc("test-03.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  // todo - error nicely?
  #[test]
  fn test_file_04() {
    let document = read_html_doc("test-04.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_05() {
    let document = read_html_doc("test-05.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_06() {
    let document = read_html_doc("test-06.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_07() {
    let document = read_html_doc("test-07.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_08() {
    let document = read_html_doc("test-08.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }
}
