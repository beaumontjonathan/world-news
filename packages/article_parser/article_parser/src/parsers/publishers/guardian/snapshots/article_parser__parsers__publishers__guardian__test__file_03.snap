---
source: src/parsers/publishers/guardian/mod.rs
expression: article
---
Ok(
    NewsArticle(
        NewsArticleContent {
            headline: "UK to ban mass gatherings in coronavirus U-turn",
            twitter_headline: None,
            description: "Measure likely to come into effect next week after sporting bodies and others act unilaterally Follow the latest updates Coronavirus: at a glance",
            thumbnail: Image {
                alt: Some(
                    "London Marathon runners",
                ),
                url: Some(
                    "https://i.guim.co.uk/img/media/eda97feb9877122b91874be75ea31599880cfc65/0_201_4256_2554/master/4256.jpg?width=300&quality=85&auto=format&fit=max&s=f8299c7c6e11ec4594042a862f9b5b90",
                ),
                caption: Some(
                    "The London Marathon is one of many sporting events in the UK to be postponed. Photograph: Jill Mead/The Guardian",
                ),
            },
            categories: [],
            images: [],
            videos: [
                Video {
                    alt: None,
                    url: Some(
                        "https://www.youtube.com/watch?v=0hG3V9378fg",
                    ),
                    caption: Some(
                        "WHO director general: Europe is now at centre of coronavirus pandemic – video",
                    ),
                },
            ],
            body: "Mass gatherings are to be banned across the UK from next weekend, the government has announced after Boris Johnson’s cautious approach to the coronavirus outbreak was overtaken by care homes, sporting bodies and even the Queen taking matters into their own hands.\nOn Thursday, despite formally moving to the delay stage of the coronavirus action plan and warning that many more families would “ lose loved ones before their time”, the prime minister stopped short of calling for mass events to be cancelled or schools to be closed. There was no specific advice for how older people should protect themselves, aside from avoiding going on cruises.\nOn Friday, the World Health Organization stressed the need for a range of measures to tackle the virus, which is thought to have infected up to 10,000 people in the UK, most of them unknowingly because they are not showing any symptoms.\nThe WHO director general, Dr Tedros Adhanom Ghebreyesus, said all possible action should be taken : “Not testing alone. Not contact tracing alone. Not quarantine alone. Not social distancing alone. Do it all.”\nHours later, in a significant change of tack, Downing Street signalled it was preparing to stop large public events, including sports fixtures and concerts, to alleviate the pressure on police and the ambulance service. It did not specify what size of event would be affected, and the timing of the clampdown has yet to be decided, but it is expected to come into force in a week’s time.\nEmergency legislation due to be passed by the government next week is expected to include compensation for organisations forced to cancel events. It is also expected to include the power to temporarily detain sufferers if necessarily.\nWhitehall sources said the country would move toward widespread home-working , although details had yet to be thrashed out.\nScores of events had already been cancelled unilaterally, as firms and institutions made their own assessments of the risk. There was a series of announcements earlier on Friday about care homes closing their doors to visitors , sporting and cultural fixtures being suspended, and universities ending in-person lectures.\nIn what is thought to be an unprecedented step, Buckingham Palace announced that the Queen would cancel public engagements in Cheshire and London next week “as a sensible precaution”. Visits from the monarch, who is 94 in April, draw crowds of well-wishers and use local council resources.\nA large number of care homes, including 120 run by Bupa, announced they would close their doors to outside visitors to protect elderly residents, despite updated government guidance suggesting that only those with virus symptoms should be turned away.\nThe Premier League, Football League, Women’s Super League and Scottish Premier League joined the global sporting shutdown . The English leagues were suspended until 3 April at least, Saturday’s Six Nations clash between Wales and Scotland was postponed, and England’s cricket series in Sri Lanka was called off. The London Marathon and The Masters in Augusta – golf’s leading tournament – were also postponed.\nIn the US, Donald Trump declared a national emergency on Friday . He told reporters afterwards that the ban on European travel to the US could be extended to include the UK. He said the exemption had been “recommended to me by a group of professionals” but he said the UK’s rising number of cases could result in a rethink. “Their numbers have gone up fairly precipitously over the last 24 hours so we may be adding them [to the ban] and we may be adding a couple of others and we may frankly start thinking about taking some off.”\nOther developments included:\nLocal and mayoral elections due to be held in May were postponed for 12 months , although the government insisted the timetable for Brexit deal negotiations would not be derailed. Hospitals were trebling the size of their intensive care units and preparing to delay millions of non-urgent procedures, including some surgery for people with heart and lung conditions. The World Health Organization stressed that countries should track and test every case of coronavirus to stop the pandemic, a day after the UK announced that only the most seriously ill in hospitals would be tested. As the WHO described Europe as the new centre of the pandemic, Britons were urged against all but essential travel to Spain and Italy; the Eiffel Tower was closed.\nJohnson has been facing increasing questions about the government’s approach, which is markedly different to that of some other countries. The former health secretary Jeremy Hunt described the UK as an “outlier”.\nOn Thursday, Johnson played down the importance of banning mass gatherings, even suggesting Scotland’s decision to ban events with more than 500 people reflected the lack of resilience of its emergency services.\n“The scientific advice, as we’ve said over the last couple of weeks, is that banning such events will have little effect on the spread,” he said.\nThe government’s chief scientific adviser, Sir Patrick Vallance, defended the approach on Friday, saying it was aimed at creating a “ herd immunity ” against Covid-19. That means allowing 60% of the population to catch the virus, but not so quickly as to overwhelm the NHS.\n“Our aim is to try and reduce the peak, broaden the peak, not suppress it completely. Also, because the vast majority of people get a mild illness, [our aim is] to build up some kind of herd immunity so more people are immune to this disease and we reduce the transmission, [while] at the same time we protect those who are most vulnerable to it. Those are the key things we need to do,” he said.\nOn Friday, the number of diagnosed cases in the UK rose to 798, though the government believes the true figure to be 5,000 to 10,000, and the first death was confirmed in Scotland.\nThe former Labour prime minister Gordon Brown has criticised the lack of a coordinated international response , urging global leaders to pledge to do “whatever it takes,” to tackle the outbreak.\nWriting in the Guardian, he likened the scale of the response necessary to the Manhattan Project, which led to the creation of the nuclear bomb. “If the Manhattan Project could bring people together in the 1940s to create the most lethal weapon in human history, surely we can come together in the 21st century to save lives and the livelihood of millions.”\nBrown blamed the spread of “populist nationalism”, which has been “subdividing our world”, for the inability of governments to deal with the crisis together. He encouraged them to be more open about how they were learning from each other’s experiences.\nHis plea came as the French president, Emmanuel Macron, announced that world leaders will hold an emergency G7 video-conference summit on Monday. “We will coordinate research efforts on a vaccine and treatments, and work on an economic and financial response,” he said.\nGlobal stock markets rallied on Friday after their worst day since the 1987 crash, on hopes that economic policymakers might deliver a coordinated response.\nIn the UK, ministers have insisted their decision-making is based on scientific advice. One factor influencing their thinking has been the risk of “fatigue” setting in if restrictive measures had to be applied for a prolonged period.\nLabour has been broadly supportive of the government’s approach so far but the shadow health secretary, Jon Ashworth, demanded more information.\n“We’re supporting to the move from the contain to delay stage, but the government needs an explain stage,” he said. “We have a duty to voice the concerns that are being raised with us: and the public want to know why the UK appears to be taking an approach which is different to other countries, when other countries themselves are saying their approach is based on their science.”\nA Whitehall source said: “Ministers are working with the chief scientific adviser and chief medical officer on our plan to stop various types of public events, including mass gatherings, beginning next week. We are also talking to businesses and other bodies about the timing of moving towards much more widespread working from home.\n“There are many complex considerations to make all these measures as effective as possible. We will make the right decisions at the right time based on the best scientific evidence.\n“For example, we are concerned about the burden large events put on public services – including the health service and the police – from dealing with coronavirus. Officials are working with industry bodies to identify how to support businesses that will be affected by this decision.\n“We have drafted emergency legislation to give the government the powers it needs to deal with coronavirus, including powers to stop mass gatherings and compensate organisations. We will publish this legislation next week.”\nParliament announced a clampdown on visitors, and urged MPs not to go on overseas visits. Members of the public would be able to watch debates from the public galleries, but not tour the parliamentary estate.\nThe Speaker, Lindsay Hoyle, said: “We are clear that now is the time to be pragmatic; everyone in the country is being asked to strike a balance and it is right that we do the same.”\n• This article was amended on 18 March 2020 because the Queen is not yet 94, as an earlier version said. She will turn 94 in April.",
            date_updated: 2020-03-18T11:16:12Z,
            date_published: 2020-03-13T22:30:11Z,
        },
    ),
)
